syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

package stock_api;


service Stock {
  rpc GetDayData(StockQuery) returns (CybosDayDatas) {}
  rpc GetMinuteData(StockQuery) returns (CybosDayDatas) {}
  rpc RequestCybosTickData(StockCodeQuery) returns (google.protobuf.Empty) {}
  rpc RequestCybosBidAsk(StockCodeQuery) returns (google.protobuf.Empty) {}
  rpc RequestCybosSubject(StockCodeQuery) returns (google.protobuf.Empty) {}

  rpc GetYesterdayTopAmountCodes(google.protobuf.Timestamp) returns (CodeList) {}

  rpc ListenCybosTickData(google.protobuf.Empty) returns (stream CybosTickData) {}
  rpc ListenCybosBidAsk(google.protobuf.Empty) returns (stream CybosBidAskTickData) {}
  rpc ListenCurrentTime(google.protobuf.Empty) returns (stream google.protobuf.Timestamp) {}
  rpc ListenCybosSubject(google.protobuf.Empty) returns (stream CybosSubjectTickData) {}

  rpc StartSimulation(SimulationArgument) returns (google.protobuf.Empty) {}
  rpc StopSimulation(google.protobuf.Empty) returns (google.protobuf.Empty) {}
}


message StockCodeQuery {
    string code = 1;
}


message CodeList {
    repeated string codelist = 1;
}


message StockQuery {
    string code = 1;
    google.protobuf.Timestamp from_datetime = 2;
    google.protobuf.Timestamp until_datetime = 3;
}


message SimulationArgument {
    google.protobuf.Timestamp from_datetime = 1;
}


message CybosDayData {
    uint32 date = 1;
    uint32 time = 2;
    uint32 start_price = 3;
    uint32 highest_price = 4; 
    uint32 lowest_price = 5;
    uint32 close_price = 6;
    uint64 volume = 7;
    uint64 amount = 8;
    uint64 cum_sell_volume = 9;
    uint64 cum_buy_volume = 10;
    int64 foreigner_hold_volume = 11;
    float foreigner_hold_rate = 12;
    int64 institution_buy_volume = 13;
    int64 institution_cum_buy_volume = 14;
}

message CybosDayDatas {
    repeated CybosDayData day_data = 1;
}


message CybosSubjectTickData {
    google.protobuf.Timestamp tick_date = 1;
    uint32 time = 2;
    string name = 3;
    string code = 4;
    string company_name = 5;
    bool buy_or_sell = 6;
    uint64 volume = 7;
    int64 total_volume = 8;
    int64 foreigner_total_volume = 9;
}


message CybosTickData {
    google.protobuf.Timestamp tick_date = 1;
    string code = 2;
    string company_name = 3;
    int32 yesterday_diff = 4;
    uint32 time = 5;
    uint32 start_price = 6;
    uint32 highest_price = 7;
    uint32 lowest_price = 8;
    uint32 ask_price = 9;
    uint32 bid_price = 10;
    uint64 cum_volume = 11;
    uint64 cum_amount = 12;
    uint32 current_price = 13;
    bool buy_or_sell = 14;
    uint64 cum_sell_volume_by_price = 15;
    uint64 cum_buy_volume_by_price = 16;
    int64 volume = 17;
    uint32 time_with_sec = 18;
    uint32 market_type_exp = 19;
    uint32 market_type = 20;
    uint64 out_time_volume = 21;
    uint64 cum_sell_volume = 22;
    uint64 cum_buy_volume = 23;
    bool is_kospi = 24;
}

message CybosBidAskTickData {
    google.protobuf.Timestamp tick_date = 1;
    string code = 2;
    uint32 time = 3;
    uint64 volume = 4;
    uint32 first_ask_price = 5;
    uint32 first_bid_price = 6;
    uint32 first_ask_remain = 7;
    uint32 first_bid_remain = 8;
    uint32 second_ask_price = 9;
    uint32 second_bid_price = 10;
    uint32 second_ask_remain = 11;
    uint32 second_bid_remain = 12;
    uint32 third_ask_price = 13;
    uint32 third_bid_price = 14;
    uint32 third_ask_remain = 15;
    uint32 third_bid_remain = 16;
    uint32 fourth_ask_price = 17;
    uint32 fourth_bid_price = 18;
    uint32 fourth_ask_remain = 19;
    uint32 fourth_bid_remain = 20;
    uint32 fifth_ask_price = 21;
    uint32 fifth_bid_price = 22;
    uint32 fifth_ask_remain = 23;
    uint32 fifth_bid_remain = 24;
    uint64 total_ask_remain = 25;
    uint64 total_bid_remain = 26;
    int64 out_time_total_ask_remain = 27;
    int64 out_time_total_bid_remain = 28;
}
